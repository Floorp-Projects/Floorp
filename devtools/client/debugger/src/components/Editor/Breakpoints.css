/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.theme-light {
  --gutter-hover-background-color: #dde1e4;
  --breakpoint-fill: var(--blue-50);
  --breakpoint-stroke: var(--blue-60);
}

.theme-dark {
  --gutter-hover-background-color: #414141;
  --breakpoint-fill: var(--blue-55);
  --breakpoint-stroke: var(--blue-40);
}

.theme-light,
.theme-dark {
  --logpoint-fill: var(--theme-graphs-purple);
  --logpoint-stroke: var(--purple-60);
  --breakpoint-condition-fill: var(--theme-graphs-yellow);
  --breakpoint-condition-stroke: var(--theme-graphs-orange);
  --breakpoint-skipped-opacity: 0.15;
  --breakpoint-inactive-opacity: 0.3;
  --breakpoint-disabled-opacity: 0.6;
}

/* Standard gutter breakpoints */
.editor-wrapper .breakpoints {
  position: absolute;
  top: 0;
  left: 0;
}

.new-breakpoint .CodeMirror-linenumber {
  pointer-events: none;
}

.editor-wrapper :not(.empty-line, .new-breakpoint)
  > .CodeMirror-gutter-wrapper
  > .CodeMirror-linenumber:hover::after {
  content: "";
  position: absolute;
  /* paint below the number */
  z-index: -1;
  top: 0;
  left: 0;
  right: -4px;
  bottom: 0;
  height: 15px;
  background-color: var(--gutter-hover-background-color);
  mask: url(chrome://devtools/content/debugger/images/breakpoint.svg)
    no-repeat;
  mask-size: auto 15px;
  mask-position: right;
}

.editor.new-breakpoint svg {
  fill: var(--breakpoint-fill);
  stroke: var(--breakpoint-stroke);
  width: 60px;
  height: 15px;
  position: absolute;
  top: 0px;
  right: -4px;
}

.editor .breakpoint {
  position: absolute;
  right: -2px;
}

.editor.new-breakpoint.folding-enabled svg {
  right: -16px;
}

.new-breakpoint.has-condition .CodeMirror-gutter-wrapper svg {
  fill: var(--breakpoint-condition-fill);
  stroke: var(--breakpoint-condition-stroke);
}

.new-breakpoint.has-log .CodeMirror-gutter-wrapper svg {
  fill: var(--logpoint-fill);
  stroke: var(--logpoint-stroke);
}

.editor.new-breakpoint.breakpoint-disabled svg {
  fill-opacity: var(--breakpoint-disabled-opacity);
  stroke-opacity: var(--breakpoint-disabled-opacity);
}

.editor-wrapper.skip-pausing .editor.new-breakpoint svg {
  fill-opacity: var(--breakpoint-skipped-opacity);
}

/* Columnn breakpoints */
.column-breakpoint {
  display: inline;
  padding-inline-start: 1px;
  padding-inline-end: 1px;
}

.column-breakpoint:hover {
  background-color: transparent;
}

.column-breakpoint svg {
  display: inline-block;
  cursor: pointer;
  height: 13px;
  width: 11px;
  vertical-align: top;
  fill: var(--breakpoint-fill);
  stroke: var(--breakpoint-stroke);
  fill-opacity: var(--breakpoint-inactive-opacity);
  stroke-opacity: var(--breakpoint-inactive-opacity);
}

.column-breakpoint.active svg {
  fill: var(--breakpoint-fill);
  stroke: var(--breakpoint-stroke);
  fill-opacity: 1;
  stroke-opacity: 1;
}

.column-breakpoint.disabled svg {
  fill-opacity: var(--breakpoint-disabled-opacity);
  stroke-opacity: var(--breakpoint-disabled-opacity);
}

.column-breakpoint.has-log.disabled svg {
  fill-opacity: 0.5;
  stroke-opacity: 0.5;
}

.column-breakpoint.has-condition svg {
  fill: var(--breakpoint-condition-fill);
  stroke: var(--breakpoint-condition-stroke);
}

.column-breakpoint.has-log svg {
  fill: var(--logpoint-fill);
  stroke: var(--logpoint-stroke);
}

.editor-wrapper.skip-pausing .column-breakpoint svg {
  fill-opacity: var(--breakpoint-skipped-opacity);
}

.img.column-marker {
  background-image: url(chrome://devtools/content/debugger/images/column-marker.svg);
}
