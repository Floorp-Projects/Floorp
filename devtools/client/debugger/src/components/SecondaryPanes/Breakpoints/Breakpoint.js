/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { PureComponent } from "react";
import { connect } from "../../../utils/connect";
import { createSelector } from "reselect";
import classnames from "classnames";
import actions from "../../../actions";
import { memoize } from "lodash";

import showContextMenu from "./BreakpointsContextMenu";
import { CloseButton } from "../../shared/Button";

import {
  getLocationWithoutColumn,
  getSelectedText,
  makeBreakpointId,
} from "../../../utils/breakpoint";
import { getSelectedLocation } from "../../../utils/selected-location";
import { features } from "../../../utils/prefs";

import {
  getBreakpointsList,
  getSelectedFrame,
  getSelectedSource,
  getCurrentThread,
  getContext,
} from "../../../selectors";

class Breakpoint extends PureComponent {
  onContextMenu = e => {
    showContextMenu({ ...this.props, contextMenuEvent: e });
  };

  get selectedLocation() {
    const { breakpoint, selectedSource } = this.props;
    return getSelectedLocation(breakpoint, selectedSource);
  }

  onDoubleClick = () => {
    const { breakpoint, openConditionalPanel } = this.props;
    if (breakpoint.options.condition) {
      openConditionalPanel(this.selectedLocation);
    } else if (breakpoint.options.logValue) {
      openConditionalPanel(this.selectedLocation, true);
    }
  };

  selectBreakpoint = event => {
    event.preventDefault();
    const { cx, selectSpecificLocation } = this.props;
    selectSpecificLocation(cx, this.selectedLocation);
  };

  removeBreakpoint = event => {
    const { cx, removeBreakpoint, breakpoint } = this.props;
    event.stopPropagation();
    removeBreakpoint(cx, breakpoint);
  };

  handleBreakpointCheckbox = () => {
    const { cx, breakpoint, enableBreakpoint, disableBreakpoint } = this.props;
    if (breakpoint.disabled) {
      enableBreakpoint(cx, breakpoint);
    } else {
      disableBreakpoint(cx, breakpoint);
    }
  };

  isCurrentlyPausedAtBreakpoint() {
    const { frame } = this.props;
    if (!frame) {
      return false;
    }

    const bpId = features.columnBreakpoints
      ? makeBreakpointId(this.selectedLocation)
      : getLocationWithoutColumn(this.selectedLocation);
    const frameId = features.columnBreakpoints
      ? makeBreakpointId(frame.selectedLocation)
      : getLocationWithoutColumn(frame.selectedLocation);
    return bpId == frameId;
  }

  getBreakpointLocation() {
    const { source } = this.props;
    const { column, line } = this.selectedLocation;

    const isWasm = source?.isWasm;
    const columnVal = features.columnBreakpoints && column ? `:${column}` : "";
    const bpLocation = isWasm
      ? `0x${line.toString(16).toUpperCase()}`
      : `${line}${columnVal}`;

    return bpLocation;
  }

  getBreakpointText() {
    const { breakpoint, selectedSource } = this.props;
    const { condition, logValue } = breakpoint.options;
    return logValue || condition || getSelectedText(breakpoint, selectedSource);
  }

  highlightText = memoize(
    (text = "", editor) => {
      const node = document.createElement("div");
      editor.CodeMirror.runMode(text, "application/javascript", node);
      return { __html: node.innerHTML };
    },
    text => text
  );

  render() {
    const { breakpoint, editor } = this.props;
    const text = this.getBreakpointText();
    const labelId = `${breakpoint.id}-label`;

    return (
      <div
        className={classnames({
          breakpoint,
          paused: this.isCurrentlyPausedAtBreakpoint(),
          disabled: breakpoint.disabled,
          "is-conditional": !!breakpoint.options.condition,
          "is-log": !!breakpoint.options.logValue,
        })}
        onClick={this.selectBreakpoint}
        onDoubleClick={this.onDoubleClick}
        onContextMenu={this.onContextMenu}
      >
        <input
          id={breakpoint.id}
          type="checkbox"
          className="breakpoint-checkbox"
          checked={!breakpoint.disabled}
          onChange={this.handleBreakpointCheckbox}
          onClick={ev => ev.stopPropagation()}
          aria-labelledby={labelId}
        />
        <span
          id={labelId}
          className="breakpoint-label cm-s-mozilla devtools-monospace"
          onClick={this.selectBreakpoint}
          title={text}
        >
          <span dangerouslySetInnerHTML={this.highlightText(text, editor)} />
        </span>
        <div className="breakpoint-line-close">
          <div className="breakpoint-line devtools-monospace">
            {this.getBreakpointLocation()}
          </div>
          <CloseButton
            handleClick={e => this.removeBreakpoint(e)}
            tooltip={L10N.getStr("breakpoints.removeBreakpointTooltip")}
          />
        </div>
      </div>
    );
  }
}

const getFormattedFrame = createSelector(
  getSelectedSource,
  getSelectedFrame,
  (selectedSource, frame) => {
    if (!frame) {
      return null;
    }

    return {
      ...frame,
      selectedLocation: getSelectedLocation(frame, selectedSource),
    };
  }
);

const mapStateToProps = (state, p) => ({
  cx: getContext(state),
  breakpoints: getBreakpointsList(state),
  frame: getFormattedFrame(state, getCurrentThread(state)),
});

export default connect(mapStateToProps, {
  enableBreakpoint: actions.enableBreakpoint,
  removeBreakpoint: actions.removeBreakpoint,
  removeBreakpoints: actions.removeBreakpoints,
  removeAllBreakpoints: actions.removeAllBreakpoints,
  disableBreakpoint: actions.disableBreakpoint,
  selectSpecificLocation: actions.selectSpecificLocation,
  setBreakpointOptions: actions.setBreakpointOptions,
  toggleAllBreakpoints: actions.toggleAllBreakpoints,
  toggleBreakpoints: actions.toggleBreakpoints,
  toggleDisabledBreakpoint: actions.toggleDisabledBreakpoint,
  openConditionalPanel: actions.openConditionalPanel,
})(Breakpoint);
