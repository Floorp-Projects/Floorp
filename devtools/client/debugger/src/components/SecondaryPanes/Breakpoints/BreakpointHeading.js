/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { PureComponent } from "react";
import { connect } from "../../../utils/connect";
import actions from "../../../actions";
import {
  getTruncatedFileName,
  getDisplayPath,
  getSourceQueryString,
  getFileURL,
} from "../../../utils/source";
import {
  getHasSiblingOfSameName,
  getBreakpointsForSource,
  getContext,
} from "../../../selectors";

import SourceIcon from "../../shared/SourceIcon";

import showContextMenu from "./BreakpointHeadingsContextMenu";

class BreakpointHeading extends PureComponent {
  onContextMenu = e => {
    showContextMenu({ ...this.props, contextMenuEvent: e });
  };

  render() {
    const {
      cx,
      sources,
      source,
      hasSiblingOfSameName,
      selectSource,
    } = this.props;

    const path = getDisplayPath(source, sources);
    const query = hasSiblingOfSameName ? getSourceQueryString(source) : "";

    return (
      <div
        className="breakpoint-heading"
        title={getFileURL(source, false)}
        onClick={() => selectSource(cx, source.id)}
        onContextMenu={this.onContextMenu}
      >
        <SourceIcon
          source={source}
          modifier={icon =>
            ["file", "javascript"].includes(icon) ? null : icon
          }
        />
        <div className="filename">
          {getTruncatedFileName(source, query)}
          {path && <span>{`../${path}/..`}</span>}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, { source }) => ({
  cx: getContext(state),
  hasSiblingOfSameName: getHasSiblingOfSameName(state, source),
  breakpointsForSource: getBreakpointsForSource(state, source.id),
});

export default connect(mapStateToProps, {
  selectSource: actions.selectSource,
  enableBreakpointsInSource: actions.enableBreakpointsInSource,
  disableBreakpointsInSource: actions.disableBreakpointsInSource,
  removeBreakpointsInSource: actions.removeBreakpointsInSource,
})(BreakpointHeading);
