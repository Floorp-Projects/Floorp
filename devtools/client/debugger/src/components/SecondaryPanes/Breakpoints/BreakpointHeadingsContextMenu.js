/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { buildMenu, showMenu } from "../../../context-menu/menu";

export default function showContextMenu(props) {
  const {
    cx,
    source,
    breakpointsForSource,
    disableBreakpointsInSource,
    enableBreakpointsInSource,
    removeBreakpointsInSource,
    contextMenuEvent,
  } = props;

  contextMenuEvent.preventDefault();

  const enableInSourceLabel = L10N.getStr(
    "breakpointHeadingsMenuItem.enableInSource.label"
  );
  const disableInSourceLabel = L10N.getStr(
    "breakpointHeadingsMenuItem.disableInSource.label"
  );
  const removeInSourceLabel = L10N.getStr(
    "breakpointHeadingsMenuItem.removeInSource.label"
  );
  const enableInSourceKey = L10N.getStr(
    "breakpointHeadingsMenuItem.enableInSource.accesskey"
  );
  const disableInSourceKey = L10N.getStr(
    "breakpointHeadingsMenuItem.disableInSource.accesskey"
  );
  const removeInSourceKey = L10N.getStr(
    "breakpointHeadingsMenuItem.removeInSource.accesskey"
  );

  const disableInSourceItem = {
    id: "node-menu-disable-in-source",
    label: disableInSourceLabel,
    accesskey: disableInSourceKey,
    disabled: false,
    click: () => disableBreakpointsInSource(cx, source),
  };

  const enableInSourceItem = {
    id: "node-menu-enable-in-source",
    label: enableInSourceLabel,
    accesskey: enableInSourceKey,
    disabled: false,
    click: () => enableBreakpointsInSource(cx, source),
  };

  const removeInSourceItem = {
    id: "node-menu-enable-in-source",
    label: removeInSourceLabel,
    accesskey: removeInSourceKey,
    disabled: false,
    click: () => removeBreakpointsInSource(cx, source),
  };

  const hideDisableInSourceItem = breakpointsForSource.every(
    breakpoint => breakpoint.disabled
  );
  const hideEnableInSourceItem = breakpointsForSource.every(
    breakpoint => !breakpoint.disabled
  );

  const items = [
    { item: disableInSourceItem, hidden: () => hideDisableInSourceItem },
    { item: enableInSourceItem, hidden: () => hideEnableInSourceItem },
    { item: removeInSourceItem, hidden: () => false },
  ];

  showMenu(contextMenuEvent, buildMenu(items));
}
