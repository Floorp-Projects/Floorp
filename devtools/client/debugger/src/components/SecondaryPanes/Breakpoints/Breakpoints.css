/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.breakpoints-pane > ._content {
  overflow-x: auto;
}

.breakpoints-toggle {
  margin: 2px 3px;
}

.breakpoints-exceptions-options *,
.breakpoints-list * {
  user-select: none;
}

.breakpoints-list {
  padding: 4px 0;
}

.breakpoints-list .breakpoint-heading {
  text-overflow: ellipsis;
  width: 100%;
  font-size: 12px;
  line-height: 16px;
}

.breakpoint-heading:not(:first-child) {
  margin-top: 2px;
}

.breakpoints-list .breakpoint-heading .filename {
  overflow: hidden;
  text-overflow: ellipsis;
}

.breakpoints-list .breakpoint-heading .filename span {
  opacity: 0.7;
  padding-left: 4px;
}

.breakpoints-list .breakpoint-heading,
.breakpoints-list .breakpoint {
  color: var(--theme-text-color-strong);
  position: relative;
  cursor: pointer;
}

.breakpoints-list .breakpoint-heading,
.breakpoints-list .breakpoint,
.breakpoints-exceptions,
.breakpoints-exceptions-caught {
  display: flex;
  align-items: center;
  overflow: hidden;
  padding-top: 2px;
  padding-bottom: 2px;
  padding-inline-start: 16px;
  padding-inline-end: 12px;
}

.breakpoints-exceptions {
  padding-bottom: 3px;
  padding-top: 3px;
  user-select: none;
}

.breakpoints-exceptions-caught {
  padding-bottom: 3px;
  padding-top: 3px;
  padding-inline-start: 36px;
}

.breakpoints-exceptions-options {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xhr-breakpoints-pane .breakpoints-exceptions-options {
  border-bottom: 1px solid var(--theme-splitter-color);
}

.breakpoints-exceptions-options:not(.empty) {
  border-bottom: 1px solid var(--theme-splitter-color);
}

.breakpoints-exceptions input,
.breakpoints-exceptions-caught input {
  padding-inline-start: 2px;
  margin-top: 0px;
  margin-bottom: 0px;
  margin-inline-start: 0;
  margin-inline-end: 2px;
  vertical-align: text-bottom;
}

.breakpoint-exceptions-label {
  line-height: 14px;
  padding-inline-end: 8px;
  cursor: default;
  overflow: hidden;
  text-overflow: ellipsis;
}

html[dir="rtl"] .breakpoints-list .breakpoint,
html[dir="rtl"] .breakpoints-list .breakpoint-heading,
html[dir="rtl"] .breakpoints-exceptions {
  border-right: 4px solid transparent;
}

html:not([dir="rtl"]) .breakpoints-list .breakpoint,
html:not([dir="rtl"]) .breakpoints-list .breakpoint-heading,
html:not([dir="rtl"]) .breakpoints-exceptions {
  border-left: 4px solid transparent;
}

html .breakpoints-list .breakpoint.is-conditional {
  border-inline-start-color: var(--theme-graphs-yellow);
}

html .breakpoints-list .breakpoint.is-log {
  border-inline-start-color: var(--theme-graphs-purple);
}

html .breakpoints-list .breakpoint.paused {
  background-color: var(--theme-toolbar-background-alt);
  border-color: var(--breakpoint-active-color);
}

.breakpoints-list .breakpoint:hover {
  background-color: var(--search-overlays-semitransparent);
}

.breakpoint-line-close {
  margin-inline-start: 4px;
}

.breakpoints-list .breakpoint .breakpoint-line {
  font-size: 11px;
  color: var(--theme-comment);
  min-width: 16px;
  text-align: end;
  padding-top: 1px;
  padding-bottom: 1px;
}

.breakpoints-list .breakpoint:hover .breakpoint-line,
.breakpoints-list .breakpoint-line-close:focus-within .breakpoint-line {
  color: transparent;
}

.breakpoints-list .breakpoint.paused:hover {
  border-color: var(--breakpoint-active-color-hover);
}

.breakpoints-list .breakpoint-label {
  display: inline-block;
  cursor: pointer;
  flex-grow: 1;
  text-overflow: ellipsis;
  overflow: hidden;
  font-size: 11px;
}

.breakpoints-list .breakpoint-label span,
.breakpoint-line-close {
  display: inline;
  line-height: 14px;
}

.breakpoint-checkbox {
  margin-inline-start: 0px;
  margin-top: 0px;
  margin-bottom: 0px;
  vertical-align: text-bottom;
}

.breakpoint-label .location {
  width: 100%;
  display: inline-block;
  overflow-x: hidden;
  text-overflow: ellipsis;
  padding: 1px 0;
  vertical-align: bottom;
}

.breakpoints-list .pause-indicator {
  flex: 0 1 content;
  order: 3;
}

.breakpoint .close-btn {
  position: absolute;
  /* hide button outside of row until hovered or focused */
  top: -100px;
}

[dir="ltr"] .breakpoint .close-btn {
  right: 12px;
}

[dir="rtl"] .breakpoint .close-btn {
  left: 12px;
}

/* Reveal the remove button on hover/focus */
.breakpoint:hover .close-btn,
.breakpoint .close-btn:focus {
  top: calc(50% - 8px);
}

/* Hide the line number when revealing the remove button (since they're overlayed) */
.breakpoint-line-close:focus-within .breakpoint-line,
.breakpoint:hover .breakpoint-line {
  visibility: hidden;
}

.CodeMirror.cm-s-mozilla-breakpoint {
  cursor: pointer;
}

.CodeMirror.cm-s-mozilla-breakpoint .CodeMirror-lines {
  padding: 0;
}

.CodeMirror.cm-s-mozilla-breakpoint .CodeMirror-sizer {
  min-width: initial !important;
}

.breakpoints-list .breakpoint .CodeMirror.cm-s-mozilla-breakpoint {
  transition: opacity 0.15s linear;
}

.breakpoints-list .breakpoint.disabled .CodeMirror.cm-s-mozilla-breakpoint {
  opacity: 0.5;
}

.CodeMirror.cm-s-mozilla-breakpoint .CodeMirror-line span[role="presentation"] {
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  display: inline-block;
}

.CodeMirror.cm-s-mozilla-breakpoint .CodeMirror-code,
.CodeMirror.cm-s-mozilla-breakpoint .CodeMirror-scroll {
  pointer-events: none;
}

.CodeMirror.cm-s-mozilla-breakpoint {
  padding-top: 1px;
}
