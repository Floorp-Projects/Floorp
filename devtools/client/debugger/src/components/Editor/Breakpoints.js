/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import Breakpoint from "./Breakpoint";

import { getSelectedSource, getFirstVisibleBreakpoints } from "../../selectors";
import { makeBreakpointId } from "../../utils/breakpoint";
import { connect } from "../../utils/connect";
import { breakpointItemActions } from "./menus/breakpoints";
import { editorItemActions } from "./menus/editor";

class Breakpoints extends Component {
  render() {
    const {
      cx,
      breakpoints,
      selectedSource,
      editor,
      breakpointActions,
      editorActions,
    } = this.props;

    if (!selectedSource || !breakpoints || selectedSource.isBlackBoxed) {
      return null;
    }

    return (
      <div>
        {breakpoints.map(bp => {
          return (
            <Breakpoint
              cx={cx}
              key={makeBreakpointId(bp.location)}
              breakpoint={bp}
              selectedSource={selectedSource}
              editor={editor}
              breakpointActions={breakpointActions}
              editorActions={editorActions}
            />
          );
        })}
      </div>
    );
  }
}

export default connect(
  state => ({
    // Retrieves only the first breakpoint per line so that the
    // breakpoint marker represents only the first breakpoint
    breakpoints: getFirstVisibleBreakpoints(state),
    selectedSource: getSelectedSource(state),
  }),
  dispatch => ({
    breakpointActions: breakpointItemActions(dispatch),
    editorActions: editorItemActions(dispatch),
  })
)(Breakpoints);
