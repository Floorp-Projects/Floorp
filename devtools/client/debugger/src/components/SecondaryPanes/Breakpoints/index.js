/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { Component } from "react";
import classnames from "classnames";
import { connect } from "../../../utils/connect";

import ExceptionOption from "./ExceptionOption";

import Breakpoint from "./Breakpoint";
import BreakpointHeading from "./BreakpointHeading";

import actions from "../../../actions";
import { getSelectedLocation } from "../../../utils/selected-location";
import { createHeadlessEditor } from "../../../utils/editor/create-editor";

import {
  makeBreakpointId,
  sortSelectedBreakpoints,
} from "../../../utils/breakpoint";

import { getSelectedSource, getBreakpointSources } from "../../../selectors";

import "./Breakpoints.css";

class Breakpoints extends Component {
  componentWillUnmount() {
    this.removeEditor();
  }

  getEditor() {
    if (!this.headlessEditor) {
      this.headlessEditor = createHeadlessEditor();
    }
    return this.headlessEditor;
  }

  removeEditor() {
    if (!this.headlessEditor) {
      return;
    }
    this.headlessEditor.destroy();
    this.headlessEditor = null;
  }

  renderExceptionsOptions() {
    const {
      breakpointSources,
      shouldPauseOnExceptions,
      shouldPauseOnCaughtExceptions,
      pauseOnExceptions,
    } = this.props;

    const isEmpty = breakpointSources.length == 0;

    return (
      <div
        className={classnames("breakpoints-exceptions-options", {
          empty: isEmpty,
        })}
      >
        <ExceptionOption
          className="breakpoints-exceptions"
          label={L10N.getStr("pauseOnExceptionsItem2")}
          isChecked={shouldPauseOnExceptions}
          onChange={() => pauseOnExceptions(!shouldPauseOnExceptions, false)}
        />

        {shouldPauseOnExceptions && (
          <ExceptionOption
            className="breakpoints-exceptions-caught"
            label={L10N.getStr("pauseOnCaughtExceptionsItem")}
            isChecked={shouldPauseOnCaughtExceptions}
            onChange={() =>
              pauseOnExceptions(true, !shouldPauseOnCaughtExceptions)
            }
          />
        )}
      </div>
    );
  }

  renderBreakpoints() {
    const { breakpointSources, selectedSource } = this.props;
    if (!breakpointSources.length) {
      return null;
    }

    const editor = this.getEditor();
    const sources = [...breakpointSources.map(({ source }) => source)];

    return (
      <div className="pane breakpoints-list">
        {breakpointSources.map(({ source, breakpoints }) => {
          const sortedBreakpoints = sortSelectedBreakpoints(
            breakpoints,
            selectedSource
          );

          return [
            <BreakpointHeading
              key={source.id}
              source={source}
              sources={sources}
            />,
            ...sortedBreakpoints.map(breakpoint => (
              <Breakpoint
                breakpoint={breakpoint}
                source={source}
                selectedSource={selectedSource}
                editor={editor}
                key={makeBreakpointId(
                  getSelectedLocation(breakpoint, selectedSource)
                )}
              />
            )),
          ];
        })}
      </div>
    );
  }

  render() {
    return (
      <div className="pane">
        {this.renderExceptionsOptions()}
        {this.renderBreakpoints()}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  breakpointSources: getBreakpointSources(state),
  selectedSource: getSelectedSource(state),
});

export default connect(mapStateToProps, {
  pauseOnExceptions: actions.pauseOnExceptions,
})(Breakpoints);
