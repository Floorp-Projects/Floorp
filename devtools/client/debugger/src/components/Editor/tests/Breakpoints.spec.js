/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import Breakpoints from "../Breakpoints";
import * as I from "immutable";

const BreakpointsComponent = Breakpoints.WrappedComponent;

function generateDefaults(overrides) {
  const sourceId = "server1.conn1.child1/source1";
  const matchingBreakpoints = { id1: { location: { sourceId } } };

  return {
    selectedSource: { sourceId, get: () => false },
    editor: {
      codeMirror: {
        setGutterMarker: jest.fn(),
      },
    },
    breakpoints: I.Map(matchingBreakpoints),
    ...overrides,
  };
}

function render(overrides = {}) {
  const props = generateDefaults(overrides);
  const component = shallow(<BreakpointsComponent {...props} />);
  return { component, props };
}

describe("Breakpoints Component", () => {
  it("should render breakpoints without columns", async () => {
    const sourceId = "server1.conn1.child1/source1";
    const breakpoints = [{ location: { sourceId } }];

    const { component, props } = render({ breakpoints });
    expect(component.find("Breakpoint")).toHaveLength(props.breakpoints.length);
  });

  it("should render breakpoints with columns", async () => {
    const sourceId = "server1.conn1.child1/source1";
    const breakpoints = [{ location: { column: 2, sourceId } }];

    const { component, props } = render({ breakpoints });
    expect(component.find("Breakpoint")).toHaveLength(props.breakpoints.length);
    expect(component).toMatchSnapshot();
  });
});
