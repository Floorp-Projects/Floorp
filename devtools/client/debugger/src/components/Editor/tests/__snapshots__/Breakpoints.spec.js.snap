// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breakpoints Component should render breakpoints with columns 1`] = `
<div>
  <Breakpoint
    breakpoint={
      Object {
        "location": Object {
          "column": 2,
          "sourceId": "server1.conn1.child1/source1",
        },
      }
    }
    editor={
      Object {
        "codeMirror": Object {
          "setGutterMarker": [MockFunction],
        },
      }
    }
    key="server1.conn1.child1/source1:undefined:2"
    selectedSource={
      Object {
        "get": [Function],
        "sourceId": "server1.conn1.child1/source1",
      }
    }
  />
</div>
`;
